### ISSUE TITLE
Cannot access net core deployed grpc service hosted in IIS: Error 500 / Details: 13 INTERNAL: Received RST_STREAM with code 0

### ISSUE THREAD

**Issue by blfuentes** [grpc/grpc-dotnet#2314]
I'm trying to follow the documentation to create a gRPC service and host it in IIS but I get response errors.
The code is almost the one provided by default.

OS: Windows 11  23H2
![image](https://github.com/grpc/grpc-dotnet/assets/5121778/23526e73-f254-4be7-81b3-657e5c4ffe43)
.net framework: net core 7.0.403
Visual Studio 2022 17.7.6
.IIS:
![image](https://github.com/grpc/grpc-dotnet/assets/5121778/5908637c-efef-4971-a4e2-9893c3080044)
grpc versions/nuget:
![image](https://github.com/grpc/grpc-dotnet/assets/5121778/4a9e2932-8645-4052-932c-35220faee4eb)

This is the **Program.cs** Server code:

```
using GrpcService.Services;

var builder = WebApplication.CreateBuilder(args);
builder.Host.UseWindowsService();
// Additional configuration is required to successfully run gRPC on macOS.
// For instructions on how to configure Kestrel and gRPC clients on macOS, visit https://go.microsoft.com/fwlink/?linkid=2099682

// Add services to the container.
builder.Services.AddGrpc();

var app = builder.Build();

app.UseGrpcWeb();

// Configure the HTTP request pipeline.
app.MapGrpcService<GreeterService>().EnableGrpcWeb();
app.MapGet("/", () => "Communication with gRPC endpoints must be made through a gRPC client. To learn how to create a client, visit: https://go.microsoft.com/fwlink/?linkid=2086909");

app.Run();
``` 
**Service**
```
public class GreeterService : Greeter.GreeterBase
{
    private readonly ILogger<GreeterService> _logger;
    public GreeterService(ILogger<GreeterService> logger)
    {
        _logger = logger;
    }

    public override Task<HelloReply> SayHello(HelloRequest request, ServerCallContext context)
    {
        return Task.FromResult(new HelloReply
        {
            Message = "Hello " + request.Name
        });
    }
}
```
**appsettings.json**
```
{
  "Logging": {
    "LogLevel": {
      "Default": "Information",
      "Microsoft.AspNetCore": "Warning"
    }
  },
  "AllowedHosts": "*",
  "Kestrel": {
    "Endpoints": {
      "Https": {
        "Url": "https://0.0.0.0:443"         
      }
    },
    "EndpointDefaults": {
      "Protocols": "Http2"
    }
  }
}
```
If I just run this in debug mode from Visual Studio, I can access to it with this client:
```
using Grpc.Net.Client;
using GrpcService.Client;

//Console.WriteLine("Waiting for server to be up...");
//await Task.Delay(5000);

using var channel = GrpcChannel.ForAddress("https://localhost:443/");
var client = new Greeter.GreeterClient(channel);
var reply = await client.SayHelloAsync(
                  new HelloRequest { Name = "GreeterClient" });
Console.WriteLine("Greeting: " + reply.Message);
Console.WriteLine("Press any key to exit...");
Console.ReadKey();
```
![image](https://github.com/grpc/grpc-dotnet/assets/5121778/c00dd201-c157-49f9-b1de-e066897a46f5)

But If I deploy it to IIS, then I just cannot access it:
using **service channel**:
![image](https://github.com/grpc/grpc-dotnet/assets/5121778/cc75ae4e-0972-47dc-aa3f-49e8df10cb6b)

using **GrpcWebHandler**
![image](https://github.com/grpc/grpc-dotnet/assets/5121778/b259a045-d715-464f-9120-6614a03ce812)

Also trying with **Postman**
![image](https://github.com/grpc/grpc-dotnet/assets/5121778/29391d65-9947-49a7-8aea-845b3085d0f7)

This is my **IIS configuration**
![image](https://github.com/grpc/grpc-dotnet/assets/5121778/84167313-dddd-4181-8057-f4010f0c9744)
![image](https://github.com/grpc/grpc-dotnet/assets/5121778/4cd70e95-a9e3-4fa4-be86-5e5b956c5a64)
(no managed code)
![image](https://github.com/grpc/grpc-dotnet/assets/5121778/f2ec5efb-3706-45a3-a1c2-a7885a94008a)

So basically I have no idea what I'm missing...

**Comment by blfuentes** [comment]
I've modified the port to use http instead of https and the error changed:

> Grpc.Core.RpcException: 'Status(StatusCode="Internal", Detail="Error starting gRPC call. HttpRequestException: An error occurred while sending the request. IOException: An HTTP/2 connection could not be established because the server did not complete the HTTP/2 handshake. HttpProtocolException: The HTTP/2 server sent invalid data on the connection. HTTP/2 error code 'PROTOCOL_ERROR' (0x1).", DebugException="System.Net.Http.HttpRequestException: An error occurred while sending the request.")'

**Comment by JamesNK** [comment]
HTTPS/TLS is required for HTTP/2 in IIS. HTTP won't work.

The error is happening in the server. The client is just reporting that it is getting a 500 error back from the server. You must look at the server logs to understand the problem.

You can simplify debugging by just browsing to `/` on the server and seeing the response message that the default template adds. Once a browser successfully calls the server on that address with `H2` negotiated then gRPC should just work.

**Comment by blfuentes** [comment]
> HTTPS/TLS is required for HTTP/2 in IIS. HTTP won't work.
> 
> The error is happening in the server. The client is just reporting that it is getting a 500 error back from the server. You must look at the server logs to understand the problem.
> 
> You can simplify debugging by just browsing to `/` on the server and seeing the response message that the default template adds. Once a browser successfully calls the server on that address with `H2` negotiated then gRPC should just work.

Thanks! I was checking the logs and setting it up with HTTPS (don't know why when I run it directly from VS in HTTP it also works if it should require HTTPS) and I was missing two modules in my IIS test server:
- https://www.iis.net/downloads/microsoft/url-rewrite
- https://dotnet.microsoft.com/en-us/download/dotnet/thank-you/runtime-aspnetcore-7.0.13-windows-hosting-bundle-installer
Once they were installed I added this basic web.config:
```
<?xml version="1.0" encoding="utf-8"?>
<configuration>
	<location path="." inheritInChildApplications="false">
		<system.webServer>
			<handlers>
				<add name="aspNetCore" path="*" verb="*" modules="AspNetCoreModuleV2" resourceType="Unspecified" />
			</handlers>
			<aspNetCore processPath="dotnet"
						arguments=".\GrpcService.dll"
						stdoutLogEnabled="false"
						stdoutLogFile=".\logs\stdout"
						hostingModel="inprocess" />
		</system.webServer>
	</location>
</configuration>
```
![image](https://github.com/grpc/grpc-dotnet/assets/5121778/10a965b9-e194-4a32-8954-f5895a68836a)
![image](https://github.com/grpc/grpc-dotnet/assets/5121778/1e2809d0-cbe6-4c5c-908d-689917a35ce8)
